text/0000-trait-alias.md

```diff
@@ -0,0 +1,270 @@
+`FooBackend` because of the universal implementation just above.
+
+## Second example: ergonomic collections and scrapping boilerplate
+
+Another example is associated types. Take the following [trait from tokio](https://docs.rs/tokio-service/0.1.0/tokio_service/trait.Service.html):
+
+```rust
+pub trait Service {
+  type Request;
+  type Response;
+  type Error;
+  type Future: Future<Item=Self::Response, Error=Self::Error>;
+  fn call(&self, req: Self::Request) -> Self::Future;
+}
+```
+
+It would be nice to be able to create a few aliases to remove boilerplate for very common
+combinations of associated types with `Service`.
+
+```rust
+Service<Request = http::Request, Response = http::Response, Error = http::Error>;
+```
+
+The trait above is a http service trait which only the associated type `Future` is left to be
+implemented. Such an alias would be very appealing because it would remove copying the whole
+`Service` trait into use sites – trait bounds, or even trait impls. Scrapping such an annoying
+boilerplate is a definitive plus to the language and might be one of the most interesting use case.
+
+# Detailed design
+[design]: #detailed-design
+
+## Syntax
+
+The syntax chosen to make a *trait alias* is:
+
+```rust
+trait TraitAlias = Trait;
+```
+
+Trait aliasing to combinations of traits is also provided with the standard `+` construct:
+
+```rust
+trait DebugDefault = Debug + Default;
+```
+
+Optionally, if needed, one can provide a `where` clause to express *bounds*:
+
+```rust
+trait DebugDefault = Debug where Self: Default; // same as the example above
+```
+
+Furthermore, it’s possible to use only the `where` clause by leaving the list of traits empty:
+
+```rust
+trait DebugDefault = where Self: Debug + Default;
+```
+
+Specifically, the grammar being added is, in informal notation:
+
+```
+ATTRIBUTE* VISIBILITY? trait IDENTIFIER(<GENERIC_PARAMS>)? = GENERIC_BOUNDS (where PREDICATES)?;
+```
+
+`GENERIC_BOUNDS` is a list of zero or more traits and lifetimes separated by `+`, the same as the current syntax for bounds on a type parameter, and `PREDICATES` is a comma-separated list of zero or more predicates, just like any other `where` clause. A trait alias containing only lifetimes (`trait Static = 'static;`) is not allowed.
+
+## Semantics
+
+Trait aliases can be used in any place arbitrary bounds would be syntactically legal.
+
+You cannot directly `impl` a trait alias, but can have them as *bounds*, *trait objects* and *impl Trait*.
+
+When using a trait alias as an object type, it is subject to object safety restrictions _after_ substituting the aliased traits. This means:
+
+1. It contains an object safe trait, optionally a lifetime, and zero or more of these other bounds: `Send`, `Sync` (that is, `trait Show = Display + Debug;` would not be object safe).
+2. All the associated types of the trait need to be specified.
+3. The `where` clause, if present, only contains bounds on `Self`.
+
+Some examples:
```

Comment:
A related question: You show some trait aliases where the RHS supplies inputs for the type parameters. And you show others where the type parameter inputs are not supplied.

Do you want to support supplying only a prefix of the inputs?

(I assume the answer is no, at least for this first version of the feature, but this should be made explicit.)


```rust
trait Inputs<X, Y> { } // I have two inputs
trait Alias1 = Inputs; // This is legal by the current RFC
trait Alias2 = Inputs<i32, i32>; // and this is also legal
trait Alias3 = Inputs<i32>; // but I assume this is not legal ...
trait Alias4<X> = Input<i32, X>; // ... and should instead be explicitly encoded like this (right?)
```